package/collection2/collection2.js

```diff
@@ -46,66 +46,66 @@ Mongo.Collection.prototype.attachSchema = function c2AttachSchema(ss, options) {
     ss = new SimpleSchema(ss);
   }
 
-  this._c2 = this._c2 || {};
-
-  // If we've already attached one schema, we combine both into a new schema unless options.replace is `true`
-  if (this._c2._simpleSchema && options.replace !== true) {
-    if (ss.version >= 2) {
-      var newSS = new SimpleSchema(this._c2._simpleSchema);
-      newSS.extend(ss);
-      ss = newSS;
-    } else {
-      ss = new SimpleSchema([this._c2._simpleSchema, ss]);
-    }
-  }
-
-  var selector = options.selector;
-
   function attachTo(obj) {
-    if (typeof selector === "object") {
+    // we need an array to hold multiple schemas
+    // position 0 is reserved for the "base" schema
+    obj._c2 = obj._c2 || {};
+    obj._c2._simpleSchemas = obj._c2._simpleSchemas || [ null ];
+
+    if (typeof options.selector === "object") {
+      // Selector Schemas
+
+      // Extend selector schema with base schema
+      var baseSchema = obj._c2._simpleSchemas[0];
+      if (baseSchema) {
+        ss = extendSchema(baseSchema.schema, ss);
+      }
+
       // Index of existing schema with identical selector
-      var schemaIndex = -1;
+      var schemaIndex;
 
-      // we need an array to hold multiple schemas
-      obj._c2._simpleSchemas = obj._c2._simpleSchemas || [];
+      // Loop through existing schemas with selectors,
+      for (schemaIndex = obj._c2._simpleSchemas.length - 1; 0 < schemaIndex; schemaIndex--) {
+        var schema = obj._c2._simpleSchemas[schemaIndex];
+        if (schema && isEqual(schema.selector, options.selector)) break;
+      }
 
-      // Loop through existing schemas with selectors
-      obj._c2._simpleSchemas.forEach((schema, index) => {
-        // if we find a schema with an identical selector, save it's index
-        if(isEqual(schema.selector, selector)) {
-          schemaIndex = index;
-        }
-      });
-      if (schemaIndex === -1) {
+      if (schemaIndex <= 0) {
         // We didn't find the schema in our array - push it into the array
         obj._c2._simpleSchemas.push({
-          schema: SimpleSchema.isSimpleSchema(ss) ? ss : new SimpleSchema(ss),
-          selector: selector,
+          schema: ss,
```

Comment:
I removed the instance initialization check because during the `c2AttachSchema` object initialization (at line 45 through 47) the ss object instance initialization check is already done, and it seemed redundant to rerun the check.

Reply:
Please also note the use of the new function at line 61 `ss = extendSchema(baseSchema.schema, ss);`

The extendSchema function is declared at line 725 and also automatically handles instance initialization.